ec2_instance: Removed nested role structure in tests

Shaps opened the pull request:
##### SUMMARY
This should make it possible for the tests to run in CI

##### ISSUE TYPE
- Test Pull Request

##### COMPONENT NAME
ec2_instance

Commit:
Removed nested role for ec2_instance tests

Also moved environment setup and environment cleanup to separate task
files

Added CI group

Updated c4.large instances to t3.nano, use ENA enabled AMI

Updated AMIs with Amazon Linux AMI

Updated us-east-1 AMI id

Specify EBS size

typo, volume_size

Removed 'state: running', updated AMI with amz ami

Also added several wait: false to speed things up

Check for instance state, correctly

Accept pending as valid state

tags_and_vpc_settings do not wait for instance to complete

wait for termination protected instance

Updated IAM role name to match with policy

Skip env cleanup when in CI

Do not wait instance in external_resource_attach

ENI remove is done in env_cleanup/CI

wait some time for instance_profile instnace to be up

Updated ebs_optimized_images

Corrected task name

Added aws_cleanup var in cloud-config-aws

typo in runme

default aws_cleanup to true

Shaps commented:
/cc @s-hertel 
I'd appreciate your feedback on this

mattclay commented:
@Shaps Now that permissions have been updated the tests are failing for a different reason:

https://app.shippable.com/github/ansible/ansible/runs/123645/102/console

> 09:56 TASK [ec2_instance : remove the VPC] *******************************************
> 10:53 An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ClientError: An error occurred (DependencyViolation) when calling the DeleteVpc operation: The vpc 'vpc-03be842359c4e9433' has dependencies and cannot be deleted.

Shaps commented:
@mattclay Thanks for looking a this. I have updated the tests to wait for the termination_protected instance and that now worked ( that was the initial issue, the issue you have pasted happens because of it when CI re-runs the tests in `-vvvv` mode ). There's now 2 new missing permissions - `ec2:AttachNetworkInterface` which causes the following 
```
02:39 An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ClientError: An error occurred (UnauthorizedOperation) when calling the AttachNetworkInterface operation: You are not authorized to perform this operation.
02:39 fatal: [localhost]: FAILED! => {"attempts": 10, "boto3_version": "1.7.15", "botocore_version": "1.12.152", "changed": false, "error": {"code": "UnauthorizedOperation", "message": "You are not authorized to perform this operation."}, "msg": "Failed to handle existing instances i-00a10fcc21eb5aff3: An error occurred (UnauthorizedOperation) when calling the AttachNetworkInterface operation: You are not authorized to perform this operation.", "resource_actions": ["ec2:DescribeInstances", "ec2:AttachNetworkInterface"], "response_metadata": {"http_headers": {"date": "Mon, 20 May 2019 22:51:30 GMT", "server": "AmazonEC2", "transfer-encoding": "chunked"}, "http_status_code": 403, "request_id": "d8b8a6d5-cc8c-4074-9e2d-073059fd2814", "retry_attempts": 0}}
02:39 ...ignoring
```
I have raised [#27](https://github.com/mattclay/aws-terminator/pull/27) to fix this. 

The next issue is when it tries to create an instance profile, for which it doesn't have permission. I'm not sure we want `iam:CreateRole` added to the policies so I have not raised a PR for that yet, let me know if you're happy with it and can either add it to the current open PR or create a new one, following the error
```
06:03 TASK [ec2_instance : Create IAM role for test] *********************************
06:03 task path: /root/.ansible/test/tmp/ec2_instance-IdPM0c-ÅÑŚÌβŁÈ/test/integration/targets/ec2_instance/tasks/iam_instance_role.yml:11
06:04 An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ClientError: An error occurred (AccessDenied) when calling the CreateRole operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=123680.102 is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::966509639900:role/shippable-123680-102-test-policy
06:04 fatal: [localhost]: FAILED! => {"changed": false, "error": {"code": "AccessDenied", "message": "User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=123680.102 is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::966509639900:role/shippable-123680-102-test-policy", "type": "Sender"}, "msg": "Unable to create role: An error occurred (AccessDenied) when calling the CreateRole operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=123680.102 is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::966509639900:role/shippable-123680-102-test-policy", "resource_actions": ["iam:CreateRole", "iam:ListPolicies", "iam:GetRole"], "response_metadata": {"http_headers": {"content-length": "478", "content-type": "text/xml", "date": "Mon, 20 May 2019 22:54:55 GMT", "x-amzn-requestid": "4986da6c-7b52-11e9-aa3b-6f9784e2e5f5"}, "http_status_code": 403, "request_id": "4986da6c-7b52-11e9-aa3b-6f9784e2e5f5", "retry_attempts": 0}}
```

mattclay commented:
@Shaps The role names need to start with `ansible-test-sts-` to match the policy:

https://github.com/mattclay/aws-terminator/blob/2e83bd2baa93ae3358ed3f425c75fd9647b15aca/aws/policy/security-services.yaml#L19

Hopefully that's the last change needed for the tests to pass.

Shaps commented:
@mattclay Unfortunately that wasn't enough, we'll also need `iam:CreateInstanceProfile`

> 05:56 TASK [ec2_instance : Create IAM role for test] *********************************
05:56 task path: /root/.ansible/test/tmp/ec2_instance-DiJJ8A-ÅÑŚÌβŁÈ/test/integration/targets/ec2_instance/tasks/iam_instance_role.yml:11
05:58 An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ClientError: An error occurred (AccessDenied) when calling the CreateInstanceProfile operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=124324.102 is not authorized to perform: iam:CreateInstanceProfile on resource: arn:aws:iam::966509639900:instance-profile/ansible-test-sts-shippable-124324-102-test-policy
05:58 fatal: [localhost]: FAILED! => {"changed": false, "error": {"code": "AccessDenied", "message": "User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=124324.102 is not authorized to perform: iam:CreateInstanceProfile on resource: arn:aws:iam::966509639900:instance-profile/ansible-test-sts-shippable-124324-102-test-policy", "type": "Sender"}, "msg": "Unable to create instance profile for role ansible-test-sts-shippable-124324-102-test-policy: An error occurred (AccessDenied) when calling the CreateInstanceProfile operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=124324.102 is not authorized to perform: iam:CreateInstanceProfile on resource: arn:aws:iam::966509639900:instance-profile/ansible-test-sts-shippable-124324-102-test-policy", "resource_actions": ["iam:AttachRolePolicy", "iam:CreateRole", "iam:ListInstanceProfilesForRole", "iam:ListAttachedRolePolicies", "iam:CreateInstanceProfile", "iam:ListPolicies", "iam:GetRole"], "response_metadata": {"http_headers": {"content-length": "518", "content-type": "text/xml", "date": "Thu, 23 May 2019 20:40:30 GMT", "x-amzn-requestid": "072ee166-377b-44f0-9743-b5cab7fdf849, 072ee166-377b-44f0-9743-b5cab7fdf849"}, "http_status_code": 403, "request_id": "072ee166-377b-44f0-9743-b5cab7fdf849", "retry_attempts": 0}}

mattclay commented:
@Shaps OK, I'll work on the additional permissions required to get the tests running.

Shaps commented:
@mattclay hopfeully this is an exhaustive list of perms that are used by the currently failing test:
```
iam:DeleteRole
iam:ListInstanceProfilesForRole
iam:DetachRolePolicy
iam:ListAttachedRolePolicies
iam:GetRole
iam:AttachRolePolicy
iam:CreateRole
iam:CreateInstanceProfile
iam:ListPolicies
```

mattclay commented:
The policy has been updated and the tests restarted.

mattclay commented:
@Shaps I noticed this task is incorrectly named (a pre-existing issue, not new to this PR):

https://github.com/ansible/ansible/blob/3ef09496a3f756c14d45b416dcd81def1b95c25d/test/integration/targets/ec2_instance/playbooks/roles/ec2_instance/tasks/checkmode_tests.yml#L159-L164

It's not using check mode (that is done by a previous task).

mattclay commented:
@Shaps Thanks for working on this, it's looking really good.

One thing I'd like to see is splitting up the tests into multiple parallel sequences of tasks. It looks like multiple independent instances are used, so tests relating to each one of those could be run in parallel to reduce the total test run time. Making those changes is probably best done in a follow-up PR though. Are you interested in working on that? If so, https://github.com/ansible/ansible/pull/43695#issuecomment-411975791 should give you an idea how to accomplish that.

Shaps commented:
@mattclay Fixed the incorrect name. 
I'm happy to work on splitting the tests, should have some time in the next couple of weeks. I agree it should be done on a follow-up PR as this one has been blocking #55712 for some time now.

Shaps commented:
Failures seems unrelated to the ec2_instance tests can we restart the tests?

mattclay commented:
I restarted the failed tests, and ec2_key failed again:

https://app.shippable.com/github/ansible/ansible/runs/125792/103/tests

I've restarted them again to see if it fails in the same spot.

mattclay commented:
The tests failed again on Shippable, but passed for me locally. I've restarted them again.

mattclay commented:
I found one issue (possibly the source of the failure):

https://github.com/ansible/ansible/blob/738879450d1a59de443f92a2dc020bf9db8102ba/test/integration/targets/ec2_instance/runme.sh#L17

Tests shouldn't invoke `virtualenv` directly as they currently are, but use our helper script instead:

https://github.com/ansible/ansible/blob/8cf6f56a0103c51e93de9e26ce1982ce4aeb7ba0/test/integration/targets/inventory_aws_conformance/runme.sh#L5

mattclay commented:
@Shaps I've opened a PR to fix the remaining tests making direct use of `virtualenv`: https://github.com/ansible/ansible/pull/57326

mattclay commented:
@Shaps I've merged those fixes. Once you rebase your PR hopefully tests will pass.

mattclay commented on test/integration/cloud-config-aws.ini.template:
```diff
@@ -16,6 +16,9 @@ aws_access_key: @ACCESS_KEY
 aws_region: @REGION
+# aws_cleanup controls whether the environment is cleaned up after tests have completed
+# This only applies to tests that have a cleanup stage
+aws_cleanup: false
```
I discussed this with @nitzmahone and he suggested to make the default behavior to run cleanup and have the [AWS cloud plugin](https://github.com/ansible/ansible/blob/devel/test/runner/lib/cloud/aws.py) disable the cleanup. That way it's harder for users to end up with unwanted resources because they forgot to enable cleanup or didn't know they needed to.

mattclay replied:
Comment the value out and note that it defaults to true.

```suggestion
# The default value is true when using this template.
# aws_cleanup: false
```

Then update the ansible-test cloud plugin to set the value if it wasn't provided by the user.

After this section:

https://github.com/ansible/ansible/blob/91d326cb68de876cda915683f6b072e5d82bd2c2/test/runner/lib/cloud/aws.py#L100

Add something like:

```python
if 'aws_cleanup' not in ansible_vars:
    ansible_vars['aws_cleanup'] = not self.managed
```

That way it defaults to true for static config (without the setting) and false for dynamic config, while still allowing the setting to be set manually.

willthames commented:
Rebased this on devel - passes all my tests locally

willthames commented:
rebuild_merge

willthames committed:
Ensure block device instance creation outputs instances

Not waiting outputs results in a format that will never
be matched by the tests

Ensure instances get tidied up

Allow ec2:ReportInstanceStatus

ec2_instance: Improve test cleanup on failing tests

Improve describe/modify attribute error handling

Address feedback on PR

willthames commented:
@s-hertel @mattclay these tests need `ec2:ReportInstanceStatus` for the `WaiterInstanceStatus*` to work

mattclay commented:
@willthames You can open a PR at https://github.com/mattclay/aws-terminator/ to update the policy file: https://github.com/mattclay/aws-terminator/blob/master/aws/policy/compute.yaml

Once that's merged I'll deploy the updated policy.

mattclay commented:
@willthames I've merged and deployed the updated policy and restarted the tests.

willthames commented:
I'll take the iam_managed_policy change out of here, I inadvertently committed it

willthames commented:
@mattclay: still seeing
> 10:35  [WARNING]: Instances i-0ff25eb708dbff329 took too long to reach state OK.
  10:35 Waiter InstanceStatusOk failed: You are not authorized to perform this
  10:35 operation.

Really not sure what else is needed, these tests pass fine under ansible-test --docker on my machine (with a locked down test account)

willthames commented:
ok, https://github.com/mattclay/aws-terminator/pull/43 raised for ec2:DescribeInstanceStatus (thanks to the awesome permissions logging in the Ansible AWS modules!)

mattclay commented:
@willthames Thanks. I've merged and deployed the policy change and restarted the tests.

s-hertel commented on test/integration/targets/ec2_instance/defaults/main.yml:
```diff
@@ -0,0 +1,23 @@
+ec2_instance_owner: 'integration-run-{{ resource_prefix }}'
+ec2_ami_image:
+  # Amazon Linux - collected 2019-05-10
+  ap-northeast-1: ami-086063d8c95bfa211
```
I wonder if we should look up the latest official minimal amazon linux image for the region rather than keeping this in sync for new regions/versions.

willthames replied:
that's not a bad idea, although I'm quite fed up of this PR now.

s-hertel replied:
No worries. These were all minor suggestions, and things we can fix up at a later point, really. Thanks for working on this, it's appreciated.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_instance.py:
```diff
@@ -1496,7 +1496,10 @@ def handle_existing(existing_matches, changed, ec2, state):
-        ec2.modify_instance_attribute(**c)
+        try:
+            AWSRetry.jittered_backoff()(ec2.modify_instance_attribute)(**c)
+        except botocore.exceptions.ClientError as e:
```
I think this can be removed. It looks like there's already exception handling where handle_existing is called.

willthames replied:
just the exception handling? The Retry logic still makes sense, I assume

s-hertel replied:
yes, agreed

s-hertel commented on test/integration/targets/ec2_instance/tasks/block_devices.yml:
```diff
@@ -57,15 +61,13 @@
   ec2_instance_info:
     filters:
       "tag:Name": "{{ resource_prefix }}-test-ebs-vols"
-      "instance-state-name": "running"
```
Was this (and the one below) removed for a reason? Since you have `wait: true` on the task creating the instance, it should be running, right?

willthames replied:
I'll put it back and repush, and if it passes tests I guess the change wasn't needed?

willthames replied:
argh, I've just realised that's the change that caused me so much pain during the testing.

Basically if the playbook fails, it gets retried, and the second time around this ec2_instance_info has the now terminated instance from the first run as the first result and the following assert completely fails.

s-hertel replied:
Ahh. Okay. Sorry for putting you through that again!

s-hertel commented on test/integration/targets/ec2_instance/tasks/iam_instance_role.yml:
```diff
@@ -10,7 +10,7 @@
     - name: Create IAM role for test
       iam_role:
-        name: "{{ resource_prefix }}-test-policy"
+        name: "ansible-test-sts-{{ resource_prefix }}-test-policy"
```
This probably was renamed so it would pass CI, but if I see a role with this name I'm going to assume it's being used for the sts_assume_role tests (since that's when I added the resource restriction). I wonder if the resource restriction should either be made more general, or if this should get its own resource restriction added. Maybe it's fine as it is. I think I hoped the role name for those tests would make it readily apparent which tests could be responsible for leaked resources, but perhaps it's as fitting here.

willthames replied:
Up to you @s-hertel. The policies for automation only allow CreateRole on `ansible-test-sts-*` - I don't mind whether that's loosened so that we can tidy these tests.

s-hertel reviewed (commented):
some minor comments, but looks good

s-hertel committed:
Ensure only running instance facts are retrieved to match the following assertion
Otherwise lingering terminated instances may be in the result
Use the instance profile arn or the role name, but not the role arn
Mark tests as unstable

jillr commented on test/integration/targets/ec2_instance/tasks/main.yml:
```diff
@@ -0,0 +1,27 @@
+#  - EC2_REGION -> AWS_REGION
+#
+
+# - include: ../../../../../setup_ec2/tasks/common.yml module_name: ec2_instance
```
```suggestion
# - include: ../../setup_ec2/tasks/common.yml module_name: ec2_instance
```
Or is this comment still relevant?

s-hertel replied:
It doesn't appear that the common tests are functional for this particular module - see https://github.com/ansible/ansible/blob/devel/test/integration/targets/setup_ec2/tasks/common.yml#L13 for example, which is not an error the ec2_instance module will give since a name isn't required.

jillr reviewed (commented):
Looks pretty good to me, and passes locally.  Thanks!

s-hertel commented:
FYI, tests are still rather unstable for CI. Can't run these locally personally before the follow up PR here https://github.com/ansible/ansible/pull/44387 so it's hard to assess *how* unstable. Unsure whether to sink more time into it or just mark them as unstable and merge. @mattclay what do you think?

mattclay commented:
@s-hertel I'd mark as unstable and then revisit once #44387 is merged.

s-hertel commented:
Thanks @Shaps, @willthames, @mattclay, and @jillr for working on this! I have opened #58650 to keep track of the unstable tests.
